endpoints/setuid.go

```diff
@@ -164,8 +164,19 @@ func NewSetUIDEndpoint(cfg *config.Configuration, syncersByBidder map[string]use
 
 		setSiteCookie := siteCookieCheck(r.UserAgent())
 
+		// Priority Ejector Set Up
+		if !isSyncerPriority(syncer.Key(), cfg.UserSync.PriorityGroups, syncersByBidder) {
```

Comment:
As written, `isSyncerPriority` returns `false` when:
1. Bidder names found in `cfg.UserSync.PriorityGroups` are found in `syncersByBidder` but do not match `syncer.Key()`, which I believe is imposible at this point of the execution flow (explained in an earlier comment).
1. No bidder name in `cfg.UserSync.PriorityGroups` is found in `syncersByBidder` (or is equal to the `bidder` URL query value, same difference).
1. `cfg.UserSync.PriorityGroups`, which is a host configuration and doesn't change from request to request is empty.

Do we really want to return an error under the scenario in point number 3? The comment in line 76 of `usersync/ejector.go` that reads `If no priority groups, then all keys in uids are non-priority` implies to me that we don't want to exit when `cfg.UserSync.PriorityGroups` is empty because we coded into `PriorityBidderEjector` a scenario where if `cfg.UserSync.PriorityGroups` is empty, we simply choose the oldest element.

However, if we do want to keep this logic as is, the implication is that an instance of Prebid Server configured with an empty `cfg.UserSync.PriorityGroups` field, will always return the `"syncer key <whatever> is not in priority groups"` error and a `400` status code in line 174 when its `/setuid` endpoint is hit. Do we want this?

My best guess is we don't. Therefore, we should validate `cfg.UserSync.PriorityGroups` in the `config` package at startup and decide to not start Prebid Server if this field is empty given that starting PBS would imply having a `/setuid` endpoint that always throws an error.

If I'm wrong, and we do mind having a `/setuid` handler that always throws an error I guess we should fail fast in order to avoid the overhead of executing lines 44 to 168 of `endpoints/setuid.go`. To fail fast, we could either:
- Modify the handler function to check for the lenght of `cfg.UserSync.PriorityGroups` right away
  ```
  43     return httprouter.Handle(func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
     +       // If `cfg.UserSync.PriorityGroups` is empty, we better fail fast and don't bother doing any unnecessary stuff
     +       if len(cfg.UserSync.PriorityGroups) == 0 {
     +           // Same as lines 169 to 174
     +           err = errors.New("syncer key " + syncer.Key() + " is not in priority groups")
     +           w.WriteHeader(http.StatusBadRequest)
     +           metricsEngine.RecordSetUid(metrics.SetUidBadRequest)
     +           so.Errors = []error{err}
     +           so.Status = http.StatusBadRequest
     +           return
     +       }
  44         so := analytics.SetUIDObject{
  45             Status: http.StatusOK,
  46             Errors: make([]error, 0),
  47         }
  48
  49         defer pbsanalytics.LogSetUIDObject(&so)
  endpoints/setuid.go
  ```
- Or assign a much smaller "fail-fast" handler function to the `/setuid` endpoint:
  ```
  39   func NewSetUIDEndpoint(cfg *config.Configuration, syncersByBidder map[string]usersync.Syncer, gdprPermsBuilder gdpr.PermissionsBuilder, tcf2CfgBuilder gdpr.TCF2ConfigBuilder, pbsanalytics analytics.PBSAnalyticsModule, accountsFetcher stored_requests.AccountFetcher, metricsEngine metrics.MetricsEngine) httprouter.Handle {
     +     // If `cfg.UserSync.PriorityGroups` is empty, use a much simpler handler function that always throws an error
     +     if len(cfg.UserSync.PriorityGroups) == 0 {
     +         return httprouter.Handle(func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
     +             // Same as lines 169 to 174
     +             err = errors.New("syncer key " + syncer.Key() + " is not in priority groups")
     +             w.WriteHeader(http.StatusBadRequest)
     +             metricsEngine.RecordSetUid(metrics.SetUidBadRequest)
     +             so.Errors = []error{err}
     +             so.Status = http.StatusBadRequest
     +             return
     +         }
     +     }
     +
  40       encoder := usersync.Base64Encoder{}
  41       decoder := usersync.Base64Decoder{}
  42  
  43       return httprouter.Handle(func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
  44           so := analytics.SetUIDObject{
  45               Status: http.StatusOK,
  46               Errors: make([]error, 0),
  47           }
  48  
  endpoints/setuid.go
  ```
Am I making sense? What do you make of this?

Reply:
Thank you for this comment, I realized that i was using `IsSyncerPriority()` not in the right way.

A syncer doesn't have to be priority, so returning an error if `IsSyncerPriority()` returns false doesn't make sense.

The way this function should actually be used is if the cookie is too full and thus we call the PriorityEjector, if there are only priority uids remaining, we need to ensure that the syncer is also a priority element, because in the case where the syncer is not a priority, we don't want to eject a priority element for it, and only in that scenario we throw an error.

So I updated the code to support this logic, where IsSyncerPriority() is called in /setuid, and the result is attached to the PriorityEjector for efficiency so that we don't keep calling `IsSyncerPriority()` during the ejection process. 

And when `getNonPriorityUids` is called within the ejector's `Choose()` method, if there is only one element returned and the syncer is not a priority element, then that means the syncer is the only non-priority element, and we return an error.